### DISCUSSION TITLE
How to use the centerline extractor?

### DISCUSSION THREAD

**Discussion by revygabor** [FAST-Imaging/FAST#193] [Q&A] [answered]
Hi!

I try to use the CenterlineExtraction class, but I can't make it work. Can you help me with that? My current code is:

```
import fast

fast.downloadTestDataIfNotExists()


importer = fast.ImageFileImporter.create(fast.Config.getTestDataPath() + 'retina.png')
centerline_extractor = fast.CenterlineExtraction.create().connect(importer)

res = centerline_extractor.runAndGetOutputData()
```
 However it results in a mesh containing 0 vertices. What am I doing wrong?

Thank you for the help!

**Comment by smistad** [comment]
Hi

Unfortunately CenterlineExtraction is for 3D only as stated here: https://fast.eriksmistad.no/classfast_1_1CenterlineExtraction.html

You can try the skeletonization instead: https://fast.eriksmistad.no/classfast_1_1Skeletonization.html
It will however, only give you skeleton image, not a contour/mesh.

Here is a working example:
```python
import fast

fast.downloadTestDataIfNotExists()

importer = fast.ImageFileImporter.create(fast.Config.getTestDataPath() + 'retina.png')
segment = fast.BinaryThresholding.create(100).connect(importer)
skeletonize = fast.Skeletonization.create().connect(segment)
renderer = fast.SegmentationRenderer.create().connect(skeletonize)
fast.SimpleWindow2D.create().connect(renderer).run()
```

It will give you this:
![image](https://github.com/smistad/FAST/assets/176325/227a6b2f-10f1-497d-a7dc-69410359a59e)

**Reply by revygabor** [reply]
Thank you for the quick answer! My original aim is to produce a centerline from a 3D binary mask. I attach a sample input.
[vessel_mask.zip](https://github.com/smistad/FAST/files/13227548/vessel_mask.zip) (0-1 unint8 values)

And my code is:
from scipy.io import loadmat
numpy_image = loadmat('vessel_mask.mat')['vessel_mask']

import fast
fast_image = fast.Image.createFromArray(numpy_image)
centerline_extractor = fast.CenterlineExtraction.create().connect(fast_image)
res = centerline_extractor.runAndGetOutputData()

However it doesn't give any result, but crashes the google colab session without any errors.
Could the colab environment or the code be the problem?

**Reply by smistad** [reply, marked as answer]
It seems the algorithm encounters an error with speed = 0 (some edge case maybe?).
I removed the check, and then it produces an output after a while:
![image](https://github.com/smistad/FAST/assets/176325/ebd1df57-60b5-4017-bf8e-9c62544e83b4)

**Reply by revygabor** [reply]
wow, thank you!
so, now I should just find the error, I guess :)

**Reply by smistad** [reply]
I removed the exit line when it encounters speed = 0, so that you can try it yourself. But you have to download and install a new wheel from here when its done: https://github.com/smistad/FAST/actions/runs/6720804256

**Reply by revygabor** [reply]
thank you very much!
